File: Cpp/NeuralNetwork.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <random>
#include <iomanip>

#include "Data.h"

class sigmoid {
public:
	double operator () (double x) {
		return 1.0 / (1.0 + exp(-x));
	}

	double derivative(double x) {
		return x * (1.0 - x);
	}
};

template<typename Activation_Func, unsigned dim = FEATURE_NR>
class NeuralNetwork {
	struct node {
		double* weights;
		const unsigned size;
		double output;
		double delta;
		double dot_product(double* x) {
			double result = 0.0;
			for (unsigned i = 0; i < size; i++)
				result += weights[i] * x[i];
			return result;
		}
	};
	double** train_x;
	unsigned* train_y;
	const unsigned input_size;
	Activation_Func func;
	std::vector<std::vector<node>> network;

	std::vector<node> create_layer(unsigned in_size, unsigned out_size, std::mt19937_64& gen) {
		std::vector<node> layer;
		layer.reserve(out_size);
		for (unsigned i = 0; i < out_size; i++) {
			double* weights = new double[dim];
			std::uniform_real_distribution<double> dist(0.0, 1.0);
			for (unsigned j = 0; j < in_size; j++)
				weights[j] = dist(gen);
			layer.push_back({ weights, in_size });
		}
		return layer;
	}

	double* forward_pass(double* x) {
		double* x_in = x;
		for (auto& layer : network) {
			double* x_out = new double[layer.size()];
			for (unsigned i = 0; i < layer.size(); i++)
				layer[i].output = x_out[i] = func(layer[i].dot_product(x_in));
			if (x_in != x)
				delete[]x_in;
			x_in = x_out;
		}
		return x_in;
		// remember to free the memory under the return value !
	}

	void backward_pass(double y_hot[]) {
		for (unsigned j = 0; j < network.front().size(); j++) {
			node& node = network.front()[j];
			node.delta = (node.output - y_hot[j]) * func.derivative(node.output);
		}
		for (int i = static_cast<int>(network.size()) - 2; i >= 0; i--) {
			for (unsigned j = 0; j < network[i].size(); j++) {
				double err = 0.0;
				for (node& n : network[i + 1])
					err += n.weights[j] * n.delta;
				network[i][j].delta = err * func.derivative(network[i][j].output);
			}
		}
	}

	void update_weights(double* x, double eta) {
		for (node& n : network[0])
			for (unsigned j = 0; j < dim; j++)
				n.weights[j] -= eta * n.delta * x[j];
		for (unsigned i = 1; i < network.size(); i++) {
			for (node& n : network[i]) {
				for (unsigned j = 0; j < network[i - 1].size(); j++)
					n.weights[j] -= eta * n.delta * network[i - 1][j].output;
			}
		}
	}

public:
	NeuralNetwork(double** x, unsigned* y, const unsigned size, std::vector<unsigned>& layers_sizes) : 
		train_x(x), train_y(y), input_size(size) {
		std::random_device rd;
		std::mt19937_64 gen(rd());
		network.push_back(create_layer(dim, layers_sizes[0], gen));
		for (unsigned i = 1; i < layers_sizes.size(); i++)
			network.push_back(create_layer(network.front().size(), layers_sizes[i], gen));
		network.push_back(create_layer(network.front().size(), CLASS_NR, gen));
	}

	~NeuralNetwork() {
		for (unsigned i = 0; i < network.size(); i++)
			for (auto& n : network[i])
				delete[]n.weights;
	}

	void fit(double eta = 0.5, unsigned iter_nr = 500) {
		for (unsigned i = 0; i < iter_nr; i++) {
			for (unsigned j = 0; j < input_size; j++) {
				double* to_delete = forward_pass(train_x[j]);
				delete[]to_delete;
				double y_hot[CLASS_NR]{};
				y_hot[train_y[j]] = 1.0;
				backward_pass(y_hot);
				update_weights(train_x[j], eta);
			}
		}
	}
	
	double check(double** test_x, unsigned* test_y, unsigned size = TEST_SIZE) {
		int guessed = 0;
		for (unsigned i = 0; i < size; i++) {
			double* probs = forward_pass(test_x[i]);
			unsigned arg_max = 0;
			for (unsigned j = 1; j < CLASS_NR; j++)
				arg_max = (probs[j] > probs[arg_max]) ? j : arg_max;
			if (arg_max == test_y[i])
				guessed += 1;
			/*for (unsigned k = 0; k < CLASS_NR; k++)
				std::cout << probs[k] << ' ';
			std::cout << '\n';*/
			delete[]probs;
		}
		return static_cast<double>(guessed) * 100.0 / static_cast<double>(size);
	}
};

int main() {
	set_holder data = prepare_data();
	data.standardize();
	std::vector<unsigned> layers{ 5 };
	NeuralNetwork<sigmoid> nn(data.train_x, data.train_y, TRAIN_SIZE, layers);
	nn.fit(0.1, 50);
	std::cout << "Fitting done" << std::endl;
	std::cout << "Accuracy: " << nn.check(data.val_x, data.val_y) << std::endl;
	data.clear();
}

File: Python/NeuraLNetwork.py
import numpy as np


class NeuralNetwork:

    def __init__(self, input_size, output_size, layers):
        self.input_dim = input_size
        self.output_dim = output_size
        self.hidden_layers = layers
        self.network = self.build_network()

    # Train network
    def train(self, X, y, eta=0.5, n_epochs=200):
        for epoch in range(n_epochs):
            for (x_, y_) in zip(X, y):
                self._forward_pass(x_) # forward pass (update node["output"])
                yhot_ = self._one_hot_encoding(y_, self.output_dim) # one-hot target
                self._backward_pass(yhot_) # backward pass error (update node["delta"])
                self._update_weights(x_, eta) # update weights (update node["weight"])

    # Predict using argmax of logits
    def predict(self, X):
        ypred = np.array([np.argmax(self._forward_pass(x_)) for x_ in X], dtype=np.int)
        return ypred

    def build_network(self):
        def _layer(in_size, out_size):
            layer = []
            for i in range(out_size):
                weights = [np.random.random() for _ in range(in_size)]
                node = {"weights": weights, # list of weights
                        "output": None, # scalar
                        "delta": None} # scalar
                layer.append(node)
            return layer

        # Stack layers (input -> hidden -> output)
        network = []
        if len(self.hidden_layers) == 0:
            network.append(_layer(self.input_dim, self.output_dim))
        else:
            network.append(_layer(self.input_dim, self.hidden_layers[0]))
            for i in range(1, len(self.hidden_layers)):
                network.append(_layer(self.hidden_layers[i-1], self.hidden_layers[i]))
            network.append(_layer(self.hidden_layers[-1], self.output_dim))

        return network

    # Forward-pass (updates node['output'])
    def _forward_pass(self, x):
        transfer = self._sigmoid
        x_in = x
        for layer in self.network:
            x_out = []
            for node in layer:
                node['output'] = transfer(self._dotprod(node['weights'], x_in))
                x_out.append(node['output'])
            x_in = x_out # set output as next input
        return x_in

    # Backward-pass (updates node['delta'], L2 loss is assumed)
    def _backward_pass(self, yhot):
        transfer_derivative = self._sigmoid_derivative # sig' = f(sig)
        n_layers = len(self.network)
        for i in reversed(range(n_layers)): # traverse backwards
            if i == n_layers - 1:
                # Difference between logits and one-hot target
                for j, node in enumerate(self.network[i]):
                    err = node['output'] - yhot[j]
                    node['delta'] = err * transfer_derivative(node['output'])
            else:
                # Weighted sum of deltas from upper layer
                for j, node in enumerate(self.network[i]):
                    err = sum([node_['weights'][j] * node_['delta'] for node_ in self.network[i+1]])
                    node['delta'] = err * transfer_derivative(node['output'])

    # Update weights (updates node['weight'])
    def _update_weights(self, x, eta):
        for i, layer in enumerate(self.network):
            # Grab input values
            if i == 0: inputs = x
            else: inputs = [node_['output'] for node_ in self.network[i-1]]
            # Update weights
            for node in layer:
                for j, input in enumerate(inputs):
                    # dw = - learning_rate * (error * transfer') * input
                    node['weights'][j] += - eta * node['delta'] * input

    # Dot product
    def _dotprod(self, a, b):
        return sum([a_ * b_ for (a_, b_) in zip(a, b)])

    # Sigmoid (activation function)
    def _sigmoid(self, x):
        return 1.0/(1.0+np.exp(-x))

    # Sigmoid derivative
    def _sigmoid_derivative(self, sigmoid):
        return sigmoid*(1.0-sigmoid)

    # One-hot encoding
    def _one_hot_encoding(self, idx, output_dim):
        x = np.zeros(output_dim, dtype=np.int)
        x[idx] = 1
        return x